src/utils/get-comment-registry.ts

```diff
@@ -0,0 +1,431 @@
+                    DeferredCommentClaimPriorityAdjustment.trailingNonSameLine;
+                deferredCommentClaims.push(commentEntry);
+            }
+            continue; // Unnecessary, but explicit
+        } else if (attachmentKey === 'leadingComments') {
+            if (currentOwnerIsFirstImport && endsMoreThanOneLineAboveOwner) {
+                debugLog?.('Found a disconnected leading comment', {
+                    comment,
+                    owner,
+                    owner_loc: owner.loc,
+                    comment_loc: comment.loc,
+                });
+
+                // This comment is probably a disconnected comment before all imports
+                deferredCommentClaims.push({
+                    ...commentEntry,
+                    needsTopOfFileOwner: true,
+                    processingPriority:
+                        commentEntry.processingPriority +
+                        DeferredCommentClaimPriorityAdjustment.leadingAboveAllImports,
+                });
+            } else {
+                if (ownerIsSpecifier) {
+                    debugLog?.(
+                        'Deferring leading specifier comment attachment',
+                        commentEntry.commentId,
+                        commentEntry.comment.value,
+                    );
+
+                    deferredCommentClaims.push({
+                        ...commentEntry,
+                        processingPriority:
+                            commentEntry.processingPriority +
+                            DeferredCommentClaimPriorityAdjustment.leadingSpecifier,
+                    });
+                } else {
+                    debugLog?.(
+                        'Attaching',
+                        attachmentKey,
+                        commentId,
+                        (owner as any)?.imported?.name,
+                        comment.value,
+                    );
+                    commentRegistry.set(commentId, commentEntry);
+                }
+            }
+            continue; // Unnecessary, but explicit
+        } else {
+            throw new Error(
+                `Unimplemented attachmentKey ${attachmentKey} for ${nodeId(
+                    owner,
+                )}`,
+            );
+        }
+    }
+};
+
+/**
+ * Utility that walks ImportDeclarations and the associated comment nodes
+ * It returns a list of CommentEntry objects that tell you which nodes comments should be associated with
+ */
+export const getCommentRegistryFromImportDeclarations = ({
+    firstImport,
+    lastImport,
+    outputNodes,
+}: {
+    /** Original declaration, not the re-sorted output-node! */
+    firstImport: ImportDeclaration;
+    /** Original declaration, not the re-sorted output-node! */
+    lastImport: ImportDeclaration;
+
+    /** Constructed Output Nodes */
+    outputNodes: ImportDeclaration[];
+}) => {
+    if ((outputNodes.length === 0 || !firstImport, !lastImport)) {
```

Comment:
I haven't seen this syntax before in an if statement, and I'm not sure I understand it.  But, from playing around, it seems that only the last expression in the sequence is considered.  So, if I have:

```ts
function test(x, y) {
    if (x, y) {
        console.log('true');
    } else {
        console.log('false');
    }
}
```

Then the only way to get `true` is to pass a truthy `y`, and it doesn't matter what `x` is.  

Can you maybe explain what the goal of this check is?

Reply:
Brain-💨 — that was just supposed to be `||` instead of javascript’s crazy & unnecessary comma operator `,` (that should be banned — eslint-rule?).

You’re right there’s not a test-case.

The way I wrote the code that calls this function, I construct the `lastImport` from the same array that I construct the `firstImport` — so unless you’re calling this helper in new code, it’s not possible to get one of the imports to be non-null while the other is null.

Reply:
Take a look at my commits, I think we don't need `lastImport` at all anymore, so I simplified it down.

Reply:
I looked at all your commits and have no objections with them! I definitely had stuff in there that was "left-overs" from previous implementation/approaches, or just no longer relevant debugging. Thanks for cleaning all that up!